File: examples/typescript/multi-format-example.ts
/**
 * Multi-Format RDF & Custom Functions Example (TypeScript/JavaScript)
 *
 * Demonstrates:
 * 1. Loading RDF data in Turtle, N-Triples, and N-Quads formats
 * 2. SPARQL queries across all formats
 * 3. Custom function registration and usage (Jena-compatible)
 * 4. Roundtrip serialization
 *
 * This example shows W3C-compliant custom SPARQL functions,
 * similar to Apache Jena's ExtensionFunctionRegistry.
 */

import { GraphDB, FunctionRegistry, Node } from 'gonnect-nano-graphdb';

async function main() {
    console.log('=== Multi-Format RDF & Custom Functions Demo ===\n');

    // =========================================================================
    // Part 1: Load Data in Different Formats
    // =========================================================================
    console.log('1. Loading data in different RDF formats...\n');

    // Same data in 3 formats
    const turtleData = `
        @prefix ex: <http://example.org/> .
        ex:Alice ex:age "30" .
        ex:Bob ex:age "25" .
        ex:Charlie ex:score "95.5" .
    `;

    const ntriplesData = `
        <http://example.org/Alice> <http://example.org/age> "30" .
        <http://example.org/Bob> <http://example.org/age> "25" .
        <http://example.org/Charlie> <http://example.org/score> "95.5" .
    `;

    const nquadsData = `
        <http://example.org/Alice> <http://example.org/age> "30" .
        <http://example.org/Bob> <http://example.org/age> "25" <http://example.org/graph1> .
        <http://example.org/Charlie> <http://example.org/score> "95.5" .
    `;

    // Create graph databases for each format
    const dbTurtle = new GraphDB();
    await dbTurtle.loadTurtle(turtleData);
    console.log(`✓ Loaded Turtle data: ${dbTurtle.size()} triples`);

    const dbNtriples = new GraphDB();
    await dbNtriples.loadNtriples(ntriplesData);
    console.log(`✓ Loaded N-Triples data: ${dbNtriples.size()} triples`);

    const dbNquads = new GraphDB();
    await dbNquads.loadNquads(nquadsData);
    console.log(`✓ Loaded N-Quads data: ${dbNquads.size()} triples\n`);

    // =========================================================================
    // Part 2: SPARQL Queries (Same query, all formats)
    // =========================================================================
    console.log('2. Running identical SPARQL queries on all formats...\n');

    const query = `
        PREFIX ex: <http://example.org/>
        SELECT ?person ?value WHERE {
            ?person ?prop ?value .
        }
    `;

    const resultsTurtle = await dbTurtle.query(query);
    const resultsNtriples = await dbNtriples.query(query);
    const resultsNquads = await dbNquads.query(query);

    console.log(`Turtle results: ${resultsTurtle.length} rows`);
    console.log(`N-Triples results: ${resultsNtriples.length} rows`);
    console.log(`N-Quads results: ${resultsNquads.length} rows`);
    console.log('✓ All formats produced results\n');

    // =========================================================================
    // Part 3: Custom SPARQL Functions (Jena-compatible)
    // =========================================================================
    console.log('3. Registering custom SPARQL functions...\n');

    // Create function registry
    const registry = new FunctionRegistry();

    // Register custom "double" function
    registry.register('http://example.org/double', (args: Node[]) => {
        if (args.length !== 1) return null;

        try {
            const value = parseFloat(args[0].literalValue());
            const result = value * 2.0;
            return Node.literalDecimal(result.toString());
        } catch (error) {
            return null;
        }
    });

    // Register custom "ageCategory" function
    registry.register('http://example.org/ageCategory', (args: Node[]) => {
        if (args.length !== 1) return null;

        try {
            const age = parseInt(args[0].literalValue());
            let category: string;

            if (age < 18) {
                category = 'minor';
            } else if (age < 65) {
                category = 'adult';
            } else {
                category = 'senior';
            }

            return Node.literalString(category);
        } catch (error) {
            return null;
        }
    });

    // Register custom "isEven" function
    registry.register('http://example.org/isEven', (args: Node[]) => {
        if (args.length !== 1) return null;

        try {
            const value = parseInt(args[0].literalValue());
            const isEven = value % 2 === 0;
            return Node.literalBoolean(isEven);
        } catch (error) {
            return null;
        }
    });

    console.log('✓ Registered 3 custom functions:');
    console.log('  - ex:double(x) - multiplies by 2');
    console.log('  - ex:ageCategory(age) - categorizes age');
    console.log('  - ex:isEven(x) - checks if even\n');

    // =========================================================================
    // Part 4: Use Custom Functions in SPARQL Queries
    // =========================================================================
    console.log('4. Querying with custom functions...\n');

    // Apply custom functions to query
    const dbWithFunctions = new GraphDB({ functionRegistry: registry });
    await dbWithFunctions.loadTurtle(turtleData);

    // Query 1: Double the ages
    const queryDouble = `
        PREFIX ex: <http://example.org/>
        SELECT ?person ?age (ex:double(?age) AS ?doubledAge)
        WHERE {
            ?person ex:age ?age .
        }
    `;

    const resultsDouble = await dbWithFunctions.query(queryDouble);
    console.log('Query: ex:double(?age)');
    for (const row of resultsDouble) {
        console.log(`  ${row.person} age ${row.age} → doubled: ${row.doubledAge}`);
    }
    console.log();

    // Query 2: Categorize ages
    const queryCategory = `
        PREFIX ex: <http://example.org/>
        SELECT ?person ?age (ex:ageCategory(?age) AS ?category)
        WHERE {
            ?person ex:age ?age .
        }
    `;

    const resultsCategory = await dbWithFunctions.query(queryCategory);
    console.log('Query: ex:ageCategory(?age)');
    for (const row of resultsCategory) {
        console.log(`  ${row.person} age ${row.age} → category: ${row.category}`);
    }
    console.log();

    // Query 3: Filter with custom function
    const queryFilter = `
        PREFIX ex: <http://example.org/>
        SELECT ?person ?age
        WHERE {
            ?person ex:age ?age .
            FILTER (ex:isEven(?age))
        }
    `;

    const resultsFilter = await dbWithFunctions.query(queryFilter);
    console.log('Query: FILTER(ex:isEven(?age))');
    for (const row of resultsFilter) {
        console.log(`  ${row.person} has even age: ${row.age}`);
    }
    console.log();

    // =========================================================================
    // Part 5: Chaining Custom Functions
    // =========================================================================
    console.log('5. Chaining custom functions...\n');

    // Register additional function for chaining
    registry.register('http://example.org/addTen', (args: Node[]) => {
        if (args.length !== 1) return null;

        try {
            const value = parseFloat(args[0].literalValue());
            const result = value + 10.0;
            return Node.literalDecimal(result.toString());
        } catch (error) {
            return null;
        }
    });

    const dbChain = new GraphDB({ functionRegistry: registry });
    await dbChain.loadTurtle(turtleData);

    // Chain functions: (age + 10) * 2
    const queryChain = `
        PREFIX ex: <http://example.org/>
        SELECT ?person ?age (ex:double(ex:addTen(?age)) AS ?transformed)
        WHERE {
            ?person ex:age ?age .
        }
    `;

    const resultsChain = await dbChain.query(queryChain);
    console.log('Query: ex:double(ex:addTen(?age))');
    for (const row of resultsChain) {
        const original = parseFloat(row.age);
        const expected = (original + 10.0) * 2.0;
        console.log(`  ${row.person} age ${row.age} → (${row.age} + 10) × 2 = ${row.transformed}`);
    }
    console.log();

    // =========================================================================
    // Part 6: Roundtrip Serialization
    // =========================================================================
    console.log('6. Roundtrip serialization (Turtle → N-Quads → Turtle)...\n');

    // Serialize to N-Quads
    const nquadsOutput = await dbTurtle.serializeNquads();
    console.log('Serialized to N-Quads:');
    console.log(nquadsOutput.substring(0, 200) + '...\n');

    // Parse back and verify
    const dbRoundtrip = new GraphDB();
    await dbRoundtrip.loadNquads(nquadsOutput);

    console.log(`✓ Roundtrip successful: ${dbRoundtrip.size()} triples preserved`);

    // Verify data is identical by running same query
    const resultsOriginal = await dbTurtle.query(query);
    const resultsRoundtrip = await dbRoundtrip.query(query);

    console.log(`✓ Query results identical: ${resultsOriginal.length} = ${resultsRoundtrip.length}\n`);

    // =========================================================================
    // Part 7: Named Graphs (N-Quads only feature)
    // =========================================================================
    console.log('7. Named graphs (N-Quads exclusive feature)...\n');

    // Query specific named graph
    const queryGraph = `
        PREFIX ex: <http://example.org/>
        SELECT ?person ?age
        WHERE {
            GRAPH <http://example.org/graph1> {
                ?person ex:age ?age .
            }
        }
    `;

    const resultsGraph = await dbNquads.query(queryGraph);
    console.log('Query GRAPH <http://example.org/graph1>:');
    console.log(`  Found ${resultsGraph.length} triples in named graph`);
    for (const row of resultsGraph) {
        console.log(`  ${row.person} age ${row.age}`);
    }
    console.log();

    // =========================================================================
    // Part 8: TypeScript-specific features
    // =========================================================================
    console.log('8. TypeScript-specific features...\n');

    // Type-safe query results
    interface PersonAge {
        person: string;
        age: string;
    }

    const typedResults = await dbTurtle.query<PersonAge>(`
        PREFIX ex: <http://example.org/>
        SELECT ?person ?age WHERE {
            ?person ex:age ?age .
        }
    `);

    console.log('Type-safe query results:');
    for (const row of typedResults) {
        // TypeScript knows row.person and row.age exist
        const age = parseInt(row.age);
        console.log(`  ${row.person}: ${age} years old`);
    }
    console.log();

    // =========================================================================
    // Summary
    // =========================================================================
    console.log('=== Summary ===');
    console.log('✓ Successfully loaded data in 3 RDF formats');
    console.log('✓ Executed SPARQL queries across all formats');
    console.log('✓ Registered and used 5 custom functions');
    console.log('✓ Demonstrated function chaining');
    console.log('✓ Performed roundtrip serialization');
    console.log('✓ Queried named graphs');
    console.log('✓ Used TypeScript type safety');
    console.log('\nAll features working correctly!');
}

// Run the example
main().catch((error) => {
    console.error('Error:', error);
    process.exit(1);
});

File: examples/kotlin/MultiFormatExample.kt
/**
 * Multi-Format RDF & Custom Functions Example (Kotlin/Android)
 *
 * Demonstrates:
 * 1. Loading RDF data in Turtle, N-Triples, and N-Quads formats
 * 2. SPARQL queries across all formats
 * 3. Custom function registration and usage (Jena-compatible)
 * 4. Roundtrip serialization
 * 5. Android-specific integration
 *
 * This example shows W3C-compliant custom SPARQL functions,
 * similar to Apache Jena's ExtensionFunctionRegistry.
 */

package com.example.graphdb

import com.zenya.gonnect.*
import kotlinx.coroutines.*

class MultiFormatExample {

    suspend fun runDemo() = coroutineScope {
        println("=== Multi-Format RDF & Custom Functions Demo ===\n")

        // =====================================================================
        // Part 1: Load Data in Different Formats
        // =====================================================================
        println("1. Loading data in different RDF formats...\n")

        // Same data in 3 formats
        val turtleData = """
            @prefix ex: <http://example.org/> .
            ex:Alice ex:age "30" .
            ex:Bob ex:age "25" .
            ex:Charlie ex:score "95.5" .
        """.trimIndent()

        val ntriplesData = """
            <http://example.org/Alice> <http://example.org/age> "30" .
            <http://example.org/Bob> <http://example.org/age> "25" .
            <http://example.org/Charlie> <http://example.org/score> "95.5" .
        """.trimIndent()

        val nquadsData = """
            <http://example.org/Alice> <http://example.org/age> "30" .
            <http://example.org/Bob> <http://example.org/age> "25" <http://example.org/graph1> .
            <http://example.org/Charlie> <http://example.org/score> "95.5" .
        """.trimIndent()

        // Create graph databases for each format
        val dbTurtle = GraphDB().apply {
            loadTurtle(turtleData)
        }
        println("✓ Loaded Turtle data: ${dbTurtle.size()} triples")

        val dbNtriples = GraphDB().apply {
            loadNtriples(ntriplesData)
        }
        println("✓ Loaded N-Triples data: ${dbNtriples.size()} triples")

        val dbNquads = GraphDB().apply {
            loadNquads(nquadsData)
        }
        println("✓ Loaded N-Quads data: ${dbNquads.size()} triples\n")

        // =====================================================================
        // Part 2: SPARQL Queries (Same query, all formats)
        // =====================================================================
        println("2. Running identical SPARQL queries on all formats...\n")

        val query = """
            PREFIX ex: <http://example.org/>
            SELECT ?person ?value WHERE {
                ?person ?prop ?value .
            }
        """.trimIndent()

        val resultsTurtle = dbTurtle.query(query)
        val resultsNtriples = dbNtriples.query(query)
        val resultsNquads = dbNquads.query(query)

        println("Turtle results: ${resultsTurtle.size} rows")
        println("N-Triples results: ${resultsNtriples.size} rows")
        println("N-Quads results: ${resultsNquads.size} rows")
        println("✓ All formats produced results\n")

        // =====================================================================
        // Part 3: Custom SPARQL Functions (Jena-compatible)
        // =====================================================================
        println("3. Registering custom SPARQL functions...\n")

        // Create function registry
        val registry = FunctionRegistry()

        // Register custom "double" function
        registry.register("http://example.org/double") { args ->
            if (args.size != 1) return@register null

            try {
                val value = args[0].literalValue().toDouble()
                val result = value * 2.0
                Node.literalDecimal(result.toString())
            } catch (e: Exception) {
                null
            }
        }

        // Register custom "ageCategory" function
        registry.register("http://example.org/ageCategory") { args ->
            if (args.size != 1) return@register null

            try {
                val age = args[0].literalValue().toInt()
                val category = when {
                    age < 18 -> "minor"
                    age < 65 -> "adult"
                    else -> "senior"
                }
                Node.literalString(category)
            } catch (e: Exception) {
                null
            }
        }

        // Register custom "isEven" function
        registry.register("http://example.org/isEven") { args ->
            if (args.size != 1) return@register null

            try {
                val value = args[0].literalValue().toInt()
                val isEven = value % 2 == 0
                Node.literalBoolean(isEven)
            } catch (e: Exception) {
                null
            }
        }

        println("✓ Registered 3 custom functions:")
        println("  - ex:double(x) - multiplies by 2")
        println("  - ex:ageCategory(age) - categorizes age")
        println("  - ex:isEven(x) - checks if even\n")

        // =====================================================================
        // Part 4: Use Custom Functions in SPARQL Queries
        // =====================================================================
        println("4. Querying with custom functions...\n")

        // Apply custom functions to query
        val dbWithFunctions = GraphDB(functionRegistry = registry).apply {
            loadTurtle(turtleData)
        }

        // Query 1: Double the ages
        val queryDouble = """
            PREFIX ex: <http://example.org/>
            SELECT ?person ?age (ex:double(?age) AS ?doubledAge)
            WHERE {
                ?person ex:age ?age .
            }
        """.trimIndent()

        val resultsDouble = dbWithFunctions.query(queryDouble)
        println("Query: ex:double(?age)")
        resultsDouble.forEach { row ->
            println("  ${row["person"]} age ${row["age"]} → doubled: ${row["doubledAge"]}")
        }
        println()

        // Query 2: Categorize ages
        val queryCategory = """
            PREFIX ex: <http://example.org/>
            SELECT ?person ?age (ex:ageCategory(?age) AS ?category)
            WHERE {
                ?person ex:age ?age .
            }
        """.trimIndent()

        val resultsCategory = dbWithFunctions.query(queryCategory)
        println("Query: ex:ageCategory(?age)")
        resultsCategory.forEach { row ->
            println("  ${row["person"]} age ${row["age"]} → category: ${row["category"]}")
        }
        println()

        // Query 3: Filter with custom function
        val queryFilter = """
            PREFIX ex: <http://example.org/>
            SELECT ?person ?age
            WHERE {
                ?person ex:age ?age .
                FILTER (ex:isEven(?age))
            }
        """.trimIndent()

        val resultsFilter = dbWithFunctions.query(queryFilter)
        println("Query: FILTER(ex:isEven(?age))")
        resultsFilter.forEach { row ->
            println("  ${row["person"]} has even age: ${row["age"]}")
        }
        println()

        // =====================================================================
        // Part 5: Chaining Custom Functions
        // =====================================================================
        println("5. Chaining custom functions...\n")

        // Register additional function for chaining
        registry.register("http://example.org/addTen") { args ->
            if (args.size != 1) return@register null

            try {
                val value = args[0].literalValue().toDouble()
                val result = value + 10.0
                Node.literalDecimal(result.toString())
            } catch (e: Exception) {
                null
            }
        }

        val dbChain = GraphDB(functionRegistry = registry).apply {
            loadTurtle(turtleData)
        }

        // Chain functions: (age + 10) * 2
        val queryChain = """
            PREFIX ex: <http://example.org/>
            SELECT ?person ?age (ex:double(ex:addTen(?age)) AS ?transformed)
            WHERE {
                ?person ex:age ?age .
            }
        """.trimIndent()

        val resultsChain = dbChain.query(queryChain)
        println("Query: ex:double(ex:addTen(?age))")
        resultsChain.forEach { row ->
            val original = row["age"]?.toDoubleOrNull() ?: 0.0
            val expected = (original + 10.0) * 2.0
            println("  ${row["person"]} age ${row["age"]} → (${row["age"]} + 10) × 2 = ${row["transformed"]}")
        }
        println()

        // =====================================================================
        // Part 6: Roundtrip Serialization
        // =====================================================================
        println("6. Roundtrip serialization (Turtle → N-Quads → Turtle)...\n")

        // Serialize to N-Quads
        val nquadsOutput = dbTurtle.serializeNquads()
        println("Serialized to N-Quads:")
        println(nquadsOutput.substring(0, 200) + "...\n")

        // Parse back and verify
        val dbRoundtrip = GraphDB().apply {
            loadNquads(nquadsOutput)
        }

        println("✓ Roundtrip successful: ${dbRoundtrip.size()} triples preserved")

        // Verify data is identical by running same query
        val resultsOriginal = dbTurtle.query(query)
        val resultsRoundtripResults = dbRoundtrip.query(query)

        println("✓ Query results identical: ${resultsOriginal.size} = ${resultsRoundtripResults.size}\n")

        // =====================================================================
        // Part 7: Named Graphs (N-Quads only feature)
        // =====================================================================
        println("7. Named graphs (N-Quads exclusive feature)...\n")

        // Query specific named graph
        val queryGraph = """
            PREFIX ex: <http://example.org/>
            SELECT ?person ?age
            WHERE {
                GRAPH <http://example.org/graph1> {
                    ?person ex:age ?age .
                }
            }
        """.trimIndent()

        val resultsGraph = dbNquads.query(queryGraph)
        println("Query GRAPH <http://example.org/graph1>:")
        println("  Found ${resultsGraph.size} triples in named graph")
        resultsGraph.forEach { row ->
            println("  ${row["person"]} age ${row["age"]}")
        }
        println()

        // =====================================================================
        // Part 8: Android-specific Integration
        // =====================================================================
        println("8. Android-specific features...\n")

        // Coroutine-based async queries
        val asyncResults = async {
            dbTurtle.query(query)
        }

        println("✓ Executed async query with Kotlin Coroutines")
        val results = asyncResults.await()
        println("  Received ${results.size} results asynchronously\n")

        // Kotlin extension functions
        fun GraphDB.queryPersonAges(): List<Map<String, String>> {
            return this.query("""
                PREFIX ex: <http://example.org/>
                SELECT ?person ?age WHERE { ?person ex:age ?age }
            """.trimIndent())
        }

        val ages = dbTurtle.queryPersonAges()
        println("✓ Used Kotlin extension function:")
        println("  Found ${ages.size} people with ages\n")

        // Kotlin data classes for type-safe results
        data class PersonAge(val person: String, val age: Int)

        val typedResults = dbTurtle.query("""
            PREFIX ex: <http://example.org/>
            SELECT ?person ?age WHERE { ?person ex:age ?age }
        """.trimIndent()).map { row ->
            PersonAge(
                person = row["person"] ?: "",
                age = row["age"]?.toIntOrNull() ?: 0
            )
        }

        println("✓ Mapped to Kotlin data classes:")
        typedResults.forEach { person ->
            println("  ${person.person}: ${person.age} years old")
        }
        println()

        // =====================================================================
        // Summary
        // =========================================================================
        println("=== Summary ===")
        println("✓ Successfully loaded data in 3 RDF formats")
        println("✓ Executed SPARQL queries across all formats")
        println("✓ Registered and used 5 custom functions")
        println("✓ Demonstrated function chaining")
        println("✓ Performed roundtrip serialization")
        println("✓ Queried named graphs")
        println("✓ Used Kotlin coroutines and extension functions")
        println("✓ Demonstrated type-safe data mapping")
        println("\nAll features working correctly!")
    }
}

// Android Activity integration example
/*
class GraphDBActivity : AppCompatActivity() {
    private lateinit var database: GraphDB

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_graphdb)

        lifecycleScope.launch {
            // Run example
            val example = MultiFormatExample()
            example.runDemo()

            // Initialize database for app use
            database = GraphDB().apply {
                // Load data from assets
                val turtleData = assets.open("data.ttl").bufferedReader().use { it.readText() }
                loadTurtle(turtleData)
            }

            // Query and display results
            val results = database.query("""
                PREFIX ex: <http://example.org/>
                SELECT ?s ?p ?o WHERE { ?s ?p ?o }
                LIMIT 10
            """.trimIndent())

            // Update UI with results
            updateUI(results)
        }
    }

    private fun updateUI(results: List<Map<String, String>>) {
        // Display results in RecyclerView or ListView
    }
}
*/

fun main() = runBlocking {
    val example = MultiFormatExample()
    example.runDemo()
}
